File: src/config/validate.ts
import type { Config, Backend, CustomFieldType } from "./schema.js";

export class ConfigValidationError extends Error {
  constructor(public readonly errors: string[]) {
    super(`Config validation failed:\n${errors.map((e) => `  - ${e}`).join("\n")}`);
    this.name = "ConfigValidationError";
  }
}

const VALID_BACKENDS: Backend[] = ["yaml-fs", "linear"];
const VALID_FIELD_TYPES: CustomFieldType[] = [
  "string",
  "number",
  "boolean",
  "enum",
];

export function validateConfig(config: Config): Config {
  const errors: string[] = [];

  if (!VALID_BACKENDS.includes(config.backend)) {
    errors.push(
      `backend: must be one of ${VALID_BACKENDS.join(", ")}, got "${config.backend}"`,
    );
  }

  if (!Array.isArray(config.tracks) || config.tracks.length === 0) {
    errors.push("tracks: must be a non-empty array of strings");
  }

  if (
    !Array.isArray(config.workflow_states) ||
    config.workflow_states.length === 0
  ) {
    errors.push("workflow_states: must be a non-empty array of strings");
  }

  if (typeof config.step_pattern !== "string" || !config.step_pattern) {
    errors.push("step_pattern: must be a non-empty string");
  }

  if (typeof config.plans_dir !== "string" || !config.plans_dir) {
    errors.push("plans_dir: must be a non-empty string");
  }

  if (
    typeof config.max_points !== "number" ||
    config.max_points < 1 ||
    !Number.isInteger(config.max_points)
  ) {
    errors.push("max_points: must be a positive integer");
  }

  if (!Array.isArray(config.pipeline?.skip)) {
    errors.push("pipeline.skip: must be an array");
  }

  for (const [i, field] of (config.issue_schema?.custom_fields ?? []).entries()) {
    const prefix = `issue_schema.custom_fields[${i}]`;
    if (!field.name) {
      errors.push(`${prefix}.name: required`);
    }
    if (!VALID_FIELD_TYPES.includes(field.type)) {
      errors.push(
        `${prefix}.type: must be one of ${VALID_FIELD_TYPES.join(", ")}`,
      );
    }
    if (field.type === "enum" && (!field.enum_values || field.enum_values.length === 0)) {
      errors.push(`${prefix}.enum_values: required when type is "enum"`);
    }
  }

  if (config.backend === "linear") {
    if (!config.linear?.api_key) {
      errors.push("linear.api_key: required when backend is linear");
    }
    if (!config.linear?.team) {
      errors.push("linear.team: required when backend is linear");
    }
  }

  if (errors.length > 0) {
    throw new ConfigValidationError(errors);
  }

  return config;
}

File: internal/config/validate.go
package config

import (
	"fmt"
	"sort"
	"time"

	"github.com/jasonraimondi/plan-bender/internal/agents"
)

func validate(cfg *Config) error {
	var errs []FieldError

	if len(cfg.Tracks) == 0 {
		errs = append(errs, FieldError{Field: "tracks", Message: "must not be empty"})
	}

	if cfg.MaxPoints < 1 {
		errs = append(errs, FieldError{Field: "max_points", Message: "must be at least 1"})
	}

	resolved, agentErrs := resolveAgents(cfg.rawAgents)
	errs = append(errs, agentErrs...)
	if len(agentErrs) == 0 {
		if len(resolved) == 0 {
			errs = append(errs, FieldError{Field: "agents", Message: "must have at least one enabled agent"})
		} else {
			cfg.Agents = resolved
		}
	}

	if cfg.Pipeline.BranchStrategy != "" && cfg.Pipeline.BranchStrategy != "integration" && cfg.Pipeline.BranchStrategy != "direct" {
		errs = append(errs, FieldError{
			Field:   "pipeline.branch_strategy",
			Message: fmt.Sprintf("must be 'integration' or 'direct', got %q", cfg.Pipeline.BranchStrategy),
		})
	}

	if cfg.Pipeline.SubprocessTimeout != "" {
		if d, err := time.ParseDuration(cfg.Pipeline.SubprocessTimeout); err != nil {
			errs = append(errs, FieldError{
				Field:   "pipeline.subprocess_timeout",
				Message: fmt.Sprintf("must be a Go duration string (e.g. \"30m\"), got %q: %v", cfg.Pipeline.SubprocessTimeout, err),
			})
		} else if d <= 0 {
			errs = append(errs, FieldError{
				Field:   "pipeline.subprocess_timeout",
				Message: fmt.Sprintf("must be positive, got %q", cfg.Pipeline.SubprocessTimeout),
			})
		}
	}

	if cfg.Linear.Enabled {
		if cfg.Linear.APIKey == "" {
			errs = append(errs, FieldError{Field: "linear.api_key", Message: "required when linear is enabled"})
		}
		if cfg.Linear.Team == "" {
			errs = append(errs, FieldError{Field: "linear.team", Message: "required when linear is enabled"})
		}
	}

	for i, cf := range cfg.IssueSchema.CustomFields {
		if cf.Name == "" {
			errs = append(errs, FieldError{
				Field:   fmt.Sprintf("issue_schema.custom_fields[%d].name", i),
				Message: "must not be empty",
			})
		}
		if cf.Type == "enum" && len(cf.EnumValues) == 0 {
			errs = append(errs, FieldError{
				Field:   fmt.Sprintf("issue_schema.custom_fields[%d].enum_values", i),
				Message: "required when type is enum",
			})
		}
	}

	if len(errs) > 0 {
		return &ConfigError{Errors: errs}
	}
	return nil
}

// resolveAgents validates all agent names against the registry (including disabled entries)
// and returns the slice of enabled agents with registry defaults merged with per-agent overrides.
// Returns field errors for unknown agent names.
func resolveAgents(rawAgents map[string]*AgentEntry) ([]ResolvedAgent, []FieldError) {
	if len(rawAgents) == 0 {
		return nil, nil
	}

	// Sort for deterministic output
	names := make([]string, 0, len(rawAgents))
	for name := range rawAgents {
		names = append(names, name)
	}
	sort.Strings(names)

	var errs []FieldError
	var resolved []ResolvedAgent

	for _, name := range names {
		entry := rawAgents[name]
		ac, err := agents.Get(name)
		if err != nil {
			errs = append(errs, FieldError{
				Field:   fmt.Sprintf("agents[%s]", name),
				Message: fmt.Sprintf("unknown agent %q", name),
			})
			continue
		}

		// Disabled agents are validated but not included in the resolved slice
		if entry == nil || !entry.Enabled {
			continue
		}

		ra := ResolvedAgent{
			Name:             name,
			ProjectDir:       ac.ProjectDir,
			UserDir:          ac.UserDir,
			Scope:            ac.Scope,
			GitignorePattern: ac.GitignorePattern,
			Extra:            entry.Options.Extra,
		}

		// Apply per-agent overrides over registry defaults
		if entry.Options.ProjectDir != nil {
			ra.ProjectDir = *entry.Options.ProjectDir
		}
		if entry.Options.UserDir != nil {
			ra.UserDir = *entry.Options.UserDir
		}
		if entry.Options.Scope != nil {
			ra.Scope = agents.Scope(*entry.Options.Scope)
		}
		if entry.Options.GitignorePattern != nil {
			ra.GitignorePattern = *entry.Options.GitignorePattern
		}

		resolved = append(resolved, ra)
	}

	return resolved, errs
}
